Mirror descriptor properties on GPU resources

toji opened the pull request:
Fixes #1498.

Starting to sketch this out to get a better feel for any problems that might come up. Only have done buffer, texture, and queryset. Thoughts so far:
  
 - General guiding principle of what to include and what to not include is "is it mutable".
   - example: `mappedAtCreation` is "mutable" in that the mapped state changes, so it's not exposed here.
   - counter example: a shader module `source` is immutable, but my inclination is that we probably wouldn't want to include it due to size?
 - Being able to reduce some of the indirection in the spec prose is nice.
 - Need to be careful about normalizing things like `GPUExtent3D`
 - Currently returning texture `size` as a dictionary, which incurs copies. Can create an `interface GPUExtent3DReadOnly` and friends if that's a problem (but I don't really think it is.)


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/pull/2494.html" title="Last updated on Jan 11, 2022, 10:30 PM UTC (89bbd17)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/2494/6481473...89bbd17.html" title="Last updated on Jan 11, 2022, 10:30 PM UTC (89bbd17)">Diff</a>

kainino0x commented on spec/index.bs:
```diff
@@ -2639,7 +2675,14 @@ namespace GPUTextureUsage {
+                    1. Set |t|.{{GPUTexture/sampleCount}} to |descriptor|.{{GPUTextureDescriptor/sampleCount}}.
+                    1. Set |t|.{{GPUTexture/dimension}} to |descriptor|.{{GPUTextureDescriptor/dimension}}.
+                    1. Set |t|.{{GPUTexture/format}} to |descriptor|.{{GPUTextureDescriptor/format}}.
+                    1. Set |t|.{{GPUTexture/usage}} to |descriptor|.{{GPUTextureDescriptor/usage}}.
```
nit: Attributes are "getters", so we can define algorithm steps to define what happens when you access them. Those steps could just reach into `this.[[descriptor]]` to get the value.

(I don't think attributes inherently have any storage behind them which is how I ended up with the language "backed by an immutable internal slot of the same name" in a few places like https://gpuweb.github.io/gpuweb/#dom-gpuuncapturederrorevent-error . So if we don't have [[descriptor]] then I guess we should probably do something like that.)

kainino0x commented on spec/index.bs:
```diff
@@ -2039,17 +2042,21 @@ interface GPUBuffer {
         The allowed usages for this {{GPUBuffer}}.
+</dl>
+
+{{GPUBuffer}} has the following internal slots:
```
nit: I think we should have an internal slot for the buffer descriptor and have the attributes return values from it (see comment below).

toji replied:
Updated the PR to do that. It certainly makes the PR itself smaller, since we don't need to update all the algorithms anymore. (I guess we still could, which may still be nice for readability.)

kainino0x reviewed (commented):
Didn't read all the details, but the API shape LGTM

Kangz commented on spec/index.bs:
```diff
@@ -2459,22 +2466,51 @@ that returns a new texture.
 <script type=idl>
 [Exposed=(Window, DedicatedWorker), SecureContext]
 interface GPUTexture {
+    readonly attribute GPUExtent3DDict size;
```
WDYT of even inlining the size, so you'd get .width .height and .depthOrArrayLayers?

toji replied:
We could do that, but if we keep it as a dict (or even a sequence or interface) then it becomes easier to pass around to other API calls. For example:

```
const depthTexture = device.createTexture({
  size: colorTexture.size,
  format: 'depth24plus',
  /* ... */
});
```

toji commented:
Also added attributes to GPUTextureView and GPUSampler. The rest of the objects feel more complicated for various reasons, so I'm inclined to leave them off for now:

 - As previously stated, it feels potentially heavyweight to include the `source` or `sourceMap` for `GPUShaderModule`, and the `hints` is an awkward structure that I don't see any benefit to repeating.
 - Deep structures like `GPURenderPipelineDescriptor` don't feel appropriate to resurface directly as attributes, and would be unwieldy to flatten.
 - Anything that surfaces another GPU object (IE: `GPUPipelineLayout`s exposing their list of `GPUBindGroupLayouts`) feels akin to [linking back to parent objects](https://github.com/gpuweb/gpuweb/issues/1497) and as such I think we should hold off on those till we have an answer for that issue.
   - `GPUExternalTexture`'s `source` also kinda fits into this bucket.

Also, the CI is complaining that I can't return a dictionary as an argument (a problem we somewhat anticipated) so I've added a `GPUExtent3DReadOnly` interface to surface the texture size with. The alternative, which I'd be OK with, is to go with @Kangz's suggestion of [flattening the extent attributes out](https://github.com/gpuweb/gpuweb/pull/2494#discussion_r781873779), though I mildly prefer keeping them grouped.

kainino0x commented on spec/index.bs:
```diff
@@ -2030,6 +2030,9 @@ that returns a new buffer in the [=buffer state/mapped=] or [=buffer state/unmap
 [Exposed=(Window, DedicatedWorker), SecureContext]
 interface GPUBuffer {
+    readonly attribute GPUSize64 size;
+    readonly attribute GPUBufferUsageFlags usage;
```
needs to be `unsigned long`; `[EnforceRange]` is not allowed on readonly attributes: https://webidl.spec.whatwg.org/#EnforceRange

toji replied:
Good catch, thanks! Updated all of them.

kainino0x commented on spec/index.bs:
```diff
@@ -2459,13 +2474,48 @@ that returns a new texture.
 [Exposed=(Window, DedicatedWorker), SecureContext]
 interface GPUTexture {
+    [SameObject] readonly attribute GPUExtent3DReadOnly size;
+    readonly attribute GPUIntegerCoordinate mipLevelCount;
+    readonly attribute GPUSize32 sampleCount;
+    readonly attribute GPUTextureDimension dimension;
+    readonly attribute GPUTextureFormat format;
+    readonly attribute GPUTextureUsageFlags usage;
```
same for GPUIntegerCoordinate, GPUSize32, and GPUTextureUsageFlags

kainino0x commented on spec/index.bs:
```diff
@@ -2683,10 +2733,49 @@ all previously submitted operations using it are complete.
+    readonly attribute GPUIntegerCoordinate baseMipLevel;
+    readonly attribute GPUIntegerCoordinate mipLevelCount;
+    readonly attribute GPUIntegerCoordinate baseArrayLayer;
+    readonly attribute GPUIntegerCoordinate arrayLayerCount;
```
same

kainino0x commented on spec/index.bs:
```diff
@@ -8770,11 +8913,26 @@ Queries {#queries}
 [Exposed=(Window, DedicatedWorker), SecureContext]
 interface GPUQuerySet {
+    readonly attribute GPUQueryType type;
+    readonly attribute GPUSize32 count;
```
same

kainino0x commented on spec/index.bs:
```diff
@@ -10258,6 +10416,15 @@ dictionary GPUExtent3DDict {
+
+    readonly attribute GPUIntegerCoordinate width;
+    readonly attribute GPUIntegerCoordinate height;
+    readonly attribute GPUIntegerCoordinate depthOrArrayLayers;
```
same

juj commented:
> General guiding principle of what to include and what to not include is "is it mutable". 

I would like to propose a slightly revised guiding principle of what to include: "with the information you are able to access, can you _manually_ serialize the resource back to CPU side and clone it again as new to GPU side in 1:1 form? (maybe on a later browser page visit)"

The _manually_ part is important, and means that there is no wish to actually have a .serialize() or a .clone() behavior (that functionality would be cool, but it would not help the introspection issue as it could be a black box opaque handle based operation), but instead: is the user is able to capture the info about the resource in a white box manner, and later call the same/original API steps to create a new resource with the same state?

WebGL and GL in general had big problems with this. As result, composability of WebGL tools and libraries suffered, and building WebGL profilers, inspectors, tracers and debuggers was a pain.

> counter example: a shader module source is immutable, but my inclination is that we probably wouldn't want to include it due to size?

Is the WebGPU spec developed explicitly from the perspective that implementations will be able to throw away the shader source text after compilation is done? If so, it would be great to have a debugging extension where the source was available if that extension was enabled. (also that extension would be great to give the platform specific intermediate and lowest level shader bytecodes. Requiring one to flip a custom browser config flag to enable that extension would be fine)

Kangz commented:
Overall the length of this discussion is why I think we need to do something extremely limited here, or postpone after v1. There are a lot of possibilities of what we do, with no obvious best solution so this topic will spend a lot of valuable standardization time that's scarce before v1.

> I would like to propose a slightly revised guiding principle of what to include: "with the information you are able to access, can you manually serialize the resource back to CPU side and clone it again as new to GPU side in 1:1 form?

What do you consider a resource? Only texture, buffer and query set or also every other WebGPU object?

> Is the WebGPU spec developed explicitly from the perspective that implementations will be able to throw away the shader source text after compilation is done? If so, it would be great to have a debugging extension where the source was available if that extension was enabled. (also that extension would be great to give the platform specific intermediate and lowest level shader bytecodes. Requiring one to flip a custom browser config flag to enable that extension would be fine)

It is a trivial shim to keep the source on the SharedModule, it would be as much work to enable as a WebGPU extension. For the intermediate representations of the shaders, that's more browser specific, but in Chromium you can do that with `--enable-dawn-features=dump_shaders`.

juj commented:
> Only texture, buffer and query set or also every other WebGPU object?

All the WebGPU objects, more or less - except maybe the few that are intended to be transient that don't span more than a single frame in lifetime.

kainino0x commented:
Mostly obsoleted by #2919
